gc/default.c

```diff
@@ -2164,9 +2164,11 @@ heap_prepare(rb_objspace_t *objspace, rb_size_pool_t *size_pool, rb_heap_t *heap
                         /* Do steps of incremental marking or lazy sweeping. */
                         gc_continue(objspace, size_pool, heap);
 
-                        if (heap->free_pages == NULL &&
-                                !heap_page_allocate_and_initialize(objspace, size_pool, heap)) {
-                            rb_bug("cannot create a new page after major GC");
+                        if (heap->free_pages == NULL) {
+                            heap_page_allocate_and_initialize_force(objspace, size_pool, heap);
```

Comment:
We definitely don't want to use `heap_page_allocate_and_initialize_force` here because it creates a new page without changing the `allocatable_slots` counter. This could cause the heap to grow larger than expected.

Instead, I think `gc_sweep_finish_size_pool` should guarantee that the size pool has at least one empty slot or set `allocatable_slots > 0`.